### PR TITLE
1965 tc specific statements

### LINKED ISSUE #1965: Handle TC specific statements according to -cob option
### Child of #224

To ensure pure Cobol source code is correctly parsed when using the `-cob` option, our parser should reject TypeCobol specific language constructs. This issue will focus on TC-specific statements.

These ANTLR grammar rules modify Cobol code elements to define their TC counterparts:
- `cobol2002Statement`
- `initializeStatement`
- `callStatement`
- `moveSimple`
- `moveCorresponding`
- `setStatementForAssignment`
- `setStatementForIndexes`
- `setStatementForConditions`

For each of these, we have to check in the builder whether it is a valid construct (in TypeCobol context) or it is forbidden (in Cobol context). To reduce the number of tests in builder, we may change the grammar and introduce intermediary constructs to factorize TC specific elements.

**To discuss:**
- how the diagnostic for invalid code elements should be created/reported ?
- how to ensure that the rest of the parsing is not affected by this invalid part ?

### PR DESCRIPTION
**Pull request by fm-117**
resolves #1965.

- `cobol2002Statement` and `initializeStatement` are not concerned as they both refer to Cobol V6 statements.
- other production rules (`callStatement`, `moveSimple`, `moveCorresponding`, `setStatementForAssignment`, `setStatementForIndexes`, `setStatementForConditions`) are now checked according to the `IsCobolLanguage` option and custom diagnostics are created if the source code is not compliant with the option.

### CODE REVIEW

**Review by rooksdo — DISMISSED** [review]

File: TypeCobol.Test/Utils/Comparator.cs [hunk 1]
```diff
@@ -208,8 +202,11 @@ public int GetTestCount()
 				}
                 foreach (var comparator in comparators) {
                     Console.WriteLine(comparator.paths.Result + " checked with " + comparator.GetType().Name);
-					var unit = new TestUnit(comparator, _copyExtensions, autoRemarks);
-                    unit.IsCobolLanguage = IsCobolLanguage;
+					var unit = new TestUnit(comparator, _copyExtensions);
```
Comment (delevoye) [inline TypeCobol.Test/Utils/Comparator.cs:205]: Inconsistent indent.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs [hunk 2]
```diff
@@ -217,10 +221,8 @@ internal AllocateStatement CreateAllocateStatement(CodeElementsParser.AllocateSt
 			}
 
 			// Register call parameters (shared storage areas) information at the CodeElement level
-			var callSite = new CallSite() { CallTarget = statement.ProgramOrProgramEntryOrProcedureOrFunction.SymbolReference };
-			int parametersCount =
-				(statement.InputParameters != null ? statement.InputParameters.Count : 0)
-				+ (statement.OutputParameter != null ? 1 : 0);
+			var callSite = new CallSite() { CallTarget = statement.ProgramOrProgramEntryOrProcedureOrFunction?.SymbolReference };
+			int parametersCount = statement.InputParameters.Count + (statement.OutputParameter != null ? 1 : 0);
```
Comment (delevoye) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs:225]: Why has the test of nullity on `statement.InputParameters` been removed here and is still present on line 226?

**Review by delevoye — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs [hunk 3]
```diff
@@ -19,18 +20,26 @@ internal partial class CodeElementBuilder: CodeElementsBaseListener {
         private ParserRuleContext Context;
 		/// <summary>CodeElement object resulting of the visit the parse tree</summary>
 		public CodeElement CodeElement { get; set; }
-		private CobolWordsBuilder CobolWordsBuilder { get; set; }
-		private CobolExpressionsBuilder CobolExpressionsBuilder { get; set; }
-		private CobolStatementsBuilder CobolStatementsBuilder { get; set; }
+        private TypeCobolOptions CompilerOptions { get; }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolCodeElementBuilder.cs]: See d419a3cb70b9cd23610e8f4db2e02978fff2274f.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 4]
```diff
@@ -29,12 +30,18 @@ internal class CobolExpressionsBuilder
         // List of program, method, or function call instructions (with shared sotrage areas)
         internal IList<CallSite> callSites { get; set; }
 
-        public CobolExpressionsBuilder(CobolWordsBuilder cobolWordsBuilder)
-		{
+        public CobolExpressionsBuilder(CobolWordsBuilder cobolWordsBuilder, TypeCobolOptions compilerOptions)
+        {
+            CompilerOptions = compilerOptions;
 			CobolWordsBuilder = cobolWordsBuilder;
+        }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs]: See d419a3cb70b9cd23610e8f4db2e02978fff2274f.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs [hunk 5]
```diff
@@ -4,21 +4,25 @@
...
+            CompilerOptions = compilerOptions;
 			CobolWordsBuilder = cobolWordsBuilder;
 			CobolExpressionsBuilder = cobolExpressionsBuilder;
 		}
 
-		private CobolWordsBuilder CobolWordsBuilder { get; set; }
-		private CobolExpressionsBuilder CobolExpressionsBuilder { get; set; }
+        private TypeCobolOptions CompilerOptions { get; }
+		private CobolWordsBuilder CobolWordsBuilder { get; }
+		private CobolExpressionsBuilder CobolExpressionsBuilder { get; }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs]: See d419a3cb70b9cd23610e8f4db2e02978fff2274f.

File: TypeCobol.Test/Utils/Comparator.cs (hunk 1 above)
Comment (fm-117) [reply TypeCobol.Test/Utils/Comparator.cs]: See d419a3cb70b9cd23610e8f4db2e02978fff2274f.

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs (hunk 2 above)
Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolStatementsBuilder.cs:225]: Nullity test removed, d419a3cb70b9cd23610e8f4db2e02978fff2274f.

**Review by delevoye — APPROVED** [review]

File: TypeCobol.Test/Misc/LanguageLevel/tcCall.rdzMix.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+       01 var-output PIC X.
+       PROCEDURE DIVISION.
+      *OK
+           CALL var-program-name USING var-input var-inout var-output.
+      *KO procedure-style CALL
+Line 13[34,38] <48, Error, Syntax> - TypeCobol-only feature: 'INPUT' keyword is not supported in a Cobol CALL.
+Line 13[12,50] <30, Error, Semantics> - Semantic error: Function not found 'var-program-name' input(Alphanumeric, ?, Alphanumeric) output(Alphanumeric)
+           CALL var-program-name INPUT var-input 
+Line 14[34,39] <30, Error, Semantics> - Semantic error: Symbol IN-OUT is not referenced
+                                 IN-OUT var-inout
```
Comment (rooksdo) [inline TypeCobol.Test/Misc/LanguageLevel/tcCall.rdzMix.txt:17]: IN-OUT is also a TypeCobol keyword but doesn't get the same message as the other INPUT/OUTPUT labels "TypeCobol-only feature: ..." ?

Comment (fm-117) [reply TypeCobol.Test/Misc/LanguageLevel/tcCall.rdzMix.txt:17]: The fact that `IN-OUT` is a _specific_ TypeCobol keyword is the reason why the message is different. In #1968, we modified the scanner so as to yield a `UserDefinedWord` for `IN-OUT` so here we get the corresponding error: `IN-OUT` is considered as one of the input arguments and as it is not defined, we have a semantic error on the use of an undefined variable.
`INPUT` and `OUTPUT` are Cobol keywords so in this PR we add a special check to create a diagnostic describing the wrong use: `INPUT` and `OUTPUT` are not allowed to be used in a CALL statement.

Comment (rooksdo) [reply TypeCobol.Test/Misc/LanguageLevel/tcCall.rdzMix.txt:17]: Ok, thanks for the explanation!

**Review by rooksdo — APPROVED** [review]

**Review by mayanje — APPROVED** [review]

**Pull request merged**